File: Day 04/Rotate_Array.py
class Solution:
    def reverse(self, arr, start, end):
        while start < end:
            arr[start], arr[end] = arr[end], arr[start]
            start += 1
            end -= 1
    
    # Function to rotate an array by d elements in counter-clockwise direction.
    def rotateArr(self, arr, d):
        n = len(arr)
        d = d % n
        self.reverse(arr, 0, d - 1)
        self.reverse(arr, d, n - 1)
        self.reverse(arr, 0, n - 1)

if __name__ == "__main__":
    arr = [1, 2, 3, 4, 5, 6, 7]
    d = 2
    sol = Solution()
    sol.rotateArr(arr, d)
    for num in arr:
        print(num, end=" ")
    print()

File: Day 04/Rotate_Array.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
  public:
    void reverse(vector<int>& arr, int start, int end){
        while(start < end){
            swap(arr[start], arr[end]);
            start++;
            end--;
        }
    }
    // Function to rotate an array by d elements in counter-clockwise direction.
    void rotateArr(vector<int>& arr, int d) {
        int n = arr.size();
        d = d % n;
        reverse(arr, 0, d-1);
        reverse(arr, d, n-1);
        reverse(arr, 0, n-1);
    }
};

int main() {
    vector<int> arr = {1, 2, 3, 4, 5, 6, 7};
    int d = 2;
    Solution sol;
    sol.rotateArr(arr, d);
    for (int num : arr) {
        cout << num << " ";
    }
    cout << endl;
    return 0;
}